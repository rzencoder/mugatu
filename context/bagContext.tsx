import { useState, useContext, createContext } from 'react'

const bagContext = createContext({})

export function ProvideBag({ children }: { children: any }): JSX.Element {
  const bag = useProvideBag()
  return <bagContext.Provider value={bag}>{children}</bagContext.Provider>
}

export const useBag = (): any => {
  return useContext(bagContext)
}

const useProvideBag = () => {
  const [bag, setBag] = useState([])

  const addToBag = (item: any) => {
    const newBag = [...bag, item]
    setBag(newBag)
  }

  const removeFromBag = (item) => {
    const updatedBag = bag.filter((el) => el.id !== item.id)
    setBag(updatedBag)
  }

  return {
    bag,
    addToBag,
    removeFromBag,
  } as const
}
