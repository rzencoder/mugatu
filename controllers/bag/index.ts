export { default as addToBag } from './addToBag'
export { default as getBag } from './getBag'
export { default as removeFromBag } from './removeFromBag'
export { default as getBagOnSignIn } from './getBagOnSignIn'
