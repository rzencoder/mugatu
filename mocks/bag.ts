export const bag = [
  {
    id: 'abc',
    name: 'Blue jeans',
    slug: 'blue-jeans',
    price: 29.99,
    rrp: 39.99,
    colour: 'blue',
    category: 'jeans',
    popular: true,
    gender: 'female',
    sizes: [
      { size: '6', stock: 10 },
      { size: '8', stock: 2 },
      { size: '10', stock: 0 },
    ],
    image: {
      url: '123',
    },
    quantity: 1,
    selectedSize: '6',
  },
  {
    id: 'def',
    name: 'Black Shirt',
    slug: 'black-shirt',
    price: 19.99,
    rrp: 24.99,
    colour: 'black',
    category: 'shirt',
    popular: false,
    gender: 'male',
    sizes: [
      { size: 's', stock: 10 },
      { size: 'm', stock: 2 },
      { size: 'l', stock: 0 },
    ],
    image: {
      url: '456',
    },
    quantity: 1,
    selectedSize: 's',
  },
]
