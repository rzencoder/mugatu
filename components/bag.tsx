import { useBag } from '@/context/bagContext'
import { useWishlist } from '@/context/wishlistContext'
import { DeleteIcon, ChevronDownIcon, InfoIcon } from '@chakra-ui/icons'
import {
  Flex,
  Box,
  Link,
  Button,
  useColorMode,
  Menu,
  MenuItem,
  MenuList,
  MenuButton,
} from '@chakra-ui/react'
import Image from 'next/image'
import NextLink from 'next/link'
import { useState, useEffect } from 'react'

const deliveryOptions = [
  {
    name: 'standard delivery',
    price: 4.95,
  },
  {
    name: 'next day delivery',
    price: 6.95,
  },
  {
    name: 'click and collect standard',
    price: 4.95,
  },
  {
    name: 'click and collect next day',
    price: 6.95,
  },
]

const Bag = () => {
  const { bag, updateBag, removeFromBag } = useBag()
  const { addToWishlist } = useWishlist()
  const { colorMode } = useColorMode()
  const [delivery, setDelivery] = useState(deliveryOptions[0])

  const getQuantityOptions = (item) => {
    const options = []
    const size = item.sizes.find((el) => el.size === item.selectedSize)
    for (let i = 1; i <= size.stock; i++) {
      if (i > 10) break
      options.push(i)
    }
    return options
  }

  const calculateSubTotal = () => {
    const total = bag.reduce((acc, cur) => {
      return cur.price * cur.quantity + acc
    }, 0)
    return total.toFixed(2)
  }

  const subtotal = calculateSubTotal()

  const displayDeliveryPrice = (price) => {
    if (price === 0) {
      return 'free'
    } else {
      return `£${price}`
    }
  }

  useEffect(() => {
    if (parseFloat(subtotal) > 50) {
      if (delivery.name === 'standard delivery' || delivery.name === 'click and collect standard') {
        if (delivery.price !== 0) {
          setDelivery({ ...delivery, price: 0 })
        }
      }
    } else {
      if (delivery.price === 0) {
        const originalPrice = deliveryOptions.find((option) => option.name === delivery.name)
        setDelivery({ ...delivery, price: originalPrice.price })
      }
    }
  }, [bag, delivery])

  if (!bag || bag.length === 0) {
    return (
      <Flex direction="column" justify="center" alignItems="center" minHeight="450px">
        <Box
          textTransform="lowercase"
          fontSize="28px"
          fontWeight="600"
          width="80%"
          textAlign="center"
          padding="0 0 30px"
        >
          Your Shopping Bag Is Empty
        </Box>
        <NextLink href="/" passHref>
          <Link>
            <Button>Continue Shopping</Button>
          </Link>
        </NextLink>
      </Flex>
    )
  }

  return (
    <Box bg={colorMode === 'light' ? '#eee' : 'mainBlack'}>
      <Flex
        direction={['column', null, 'row']}
        maxWidth="1000px"
        minHeight="500px"
        margin="0 auto"
        padding={['15px 5px', '20px']}
        justifyContent="space-between"
        position="relative"
      >
        <Flex width={['100%', null, '60%']} direction="column" mr={['5px', '20px']}>
          <Flex
            direction="column"
            p={['15px 10px', '20px']}
            mb="20px"
            bg={colorMode === 'light' ? 'mainWhite' : '#111'}
          >
            <Box p="0 0 15px" fontWeight="600" fontSize="28px">
              my bag
            </Box>
            {bag.map((item) => {
              return (
                <Flex key={`bag-item-${item.id}`} p="10px 0">
                  <Flex
                    minWidth={['120px', '140px']}
                    maxWidth={['120px', '140px']}
                    maxHeight={['180px', 'none']}
                  >
                    <Image src={item.image.url} width={150} height={225}></Image>
                  </Flex>
                  <Flex
                    direction="column"
                    p={['0 0 0 10px', '0 10px 0 15px']}
                    justifyContent="space-between"
                    width="100%"
                  >
                    <Flex justifyContent="space-between">
                      <NextLink
                        href={`/${item.gender === 'female' ? 'women' : 'men'}/catalog/${item.slug}`}
                        passHref
                      >
                        <Link>
                          <Box fontSize={['16px', '18px']} fontWeight="600">
                            {item.name}
                          </Box>
                        </Link>
                      </NextLink>
                      <Button
                        variant="icon"
                        fontSize="22px"
                        padding="5px"
                        margin={['0 3px', '0 10px']}
                        onClick={() => removeFromBag(item)}
                        color={colorMode === 'light' ? '#999' : '#aaa'}
                      >
                        <DeleteIcon
                          _hover={{
                            color: colorMode === 'light' ? '#777' : '#ccc',
                          }}
                        />
                      </Button>
                    </Flex>
                    <Flex p="5px 0" fontSize={['14px', '15px']} direction="column">
                      <Box p="0">{item.colour}</Box>
                      <Box p="0">size: {item.selectedSize}</Box>
                      <Flex alignItems="center">
                        <Box>qty:</Box>
                        <Menu>
                          <MenuButton
                            fontSize="15px"
                            height="20px"
                            as={Button}
                            rightIcon={<ChevronDownIcon m="0" />}
                            variant="transparentBg"
                            textTransform="lowercase"
                          >
                            {item.quantity}
                          </MenuButton>
                          <MenuList
                            bg={colorMode === 'light' ? 'mainWhite' : 'mainBlack'}
                            fontSize="16px"
                            minWidth="0"
                          >
                            {getQuantityOptions(item).map((el) => {
                              return (
                                <MenuItem
                                  key={`product-quantity-${el}`}
                                  onClick={() => updateBag({ ...item, quantity: el })}
                                >
                                  {el}
                                </MenuItem>
                              )
                            })}
                          </MenuList>
                        </Menu>
                      </Flex>
                    </Flex>
                    <Flex p="5px 0">
                      <Box fontSize={['19px', '22px']} color="mainRed">
                        £{item.price}
                      </Box>
                      <Box
                        fontSize={['14px', '15px']}
                        textDecoration="line-through"
                        margin="3px 0 0 8px"
                        color={colorMode === 'light' ? '#333' : '#ddd'}
                      >
                        £{item.rrp}
                      </Box>
                    </Flex>
                    <Button
                      fontSize={['14px', '15px']}
                      borderWidth="1px"
                      width={['150px', '170px']}
                      height={['35px', '45px']}
                      p={['1px', '3px 5px']}
                      _hover={{ bg: colorMode === 'light' ? 'mainBlackHover' : 'mainWhiteHover' }}
                      onClick={() => {
                        addToWishlist(item)
                        removeFromBag(item)
                      }}
                    >
                      <Box
                        backgroundImage="url(/icons/heart.png)"
                        width={['15px', '20px']}
                        height={['15px', '20px']}
                        backgroundSize="cover"
                        filter={colorMode === 'light' ? 'invert()' : 'none'}
                      />
                      <Box p="0 0 2px 5px" textTransform="lowercase">
                        Move to Wishlist
                      </Box>
                    </Button>
                  </Flex>
                </Flex>
              )
            })}
          </Flex>
          <Flex
            display={['none', null, 'flex']}
            justifyContent="space-between"
            bg={colorMode === 'light' ? 'mainWhite' : '#111'}
            p="20px"
            fontSize="22px"
            fontWeight="600"
          >
            <Box>sub-total</Box>
            <Box>£{subtotal}</Box>
          </Flex>
        </Flex>
        <Flex
          width={['100%', null, '40%']}
          direction="column"
          height={['350px']}
          bg={colorMode === 'light' ? 'mainWhite' : '#111'}
          p="20px"
          justifyContent="space-between"
        >
          <Box p="0 0 15px" fontWeight="600" fontSize="28px">
            summary
          </Box>
          <Flex direction="column" fontSize="16px">
            <Flex justifyContent="space-between" fontWeight="600">
              <Box>sub-total</Box>
              <Box>£{subtotal}</Box>
            </Flex>
            {subtotal < 50 && (
              <Flex
                bg={colorMode === 'light' ? 'mainBlack' : 'mainWhite'}
                color={colorMode === 'light' ? 'mainWhite' : 'mainBlack'}
                p="8px 12px"
                alignItems="center"
                margin="10px 0"
                fontWeight="600"
              >
                <InfoIcon fontSize="22px" mr="15px" />
                <Box>{`spend £${(50 - parseFloat(subtotal)).toFixed(
                  2
                )} more to qualify for free delivery`}</Box>
              </Flex>
            )}
            <Flex direction="column">
              <Flex justifyContent="space-between" fontWeight="600">
                <Box>delivery</Box>
                <Box>{displayDeliveryPrice(delivery.price)}</Box>
              </Flex>
              <Menu>
                <MenuButton
                  textAlign="left"
                  as={Button}
                  border="none"
                  textTransform="lowercase"
                  p="0"
                  fontWeight="400"
                  fontSize="16px"
                >
                  <Box as="span">{`${delivery.name} (${displayDeliveryPrice(
                    delivery.price
                  )})`}</Box>
                  <ChevronDownIcon />
                </MenuButton>
                <MenuList>
                  {deliveryOptions.map((option, index) => {
                    const price = option.name === delivery.name ? delivery.price : option.price
                    return (
                      <MenuItem
                        key={`delivery-option-${option.name}`}
                        onClick={() => setDelivery(deliveryOptions[index])}
                      >
                        {`${option.name} (${displayDeliveryPrice(price)})`}
                      </MenuItem>
                    )
                  })}
                </MenuList>
              </Menu>
            </Flex>
            <Flex justifyContent="space-between" fontSize="22px" fontWeight="600" p="5px 0">
              <Box>total</Box>
              <Box>£{(parseFloat(subtotal) + delivery.price).toFixed(2)}</Box>
            </Flex>
          </Flex>
          <NextLink href="/checkout" passHref>
            <Link>
              <Button
                bg="#0d6d33"
                border="none"
                color="mainWhite"
                width="100%"
                _hover={{
                  bg: '#15ab51',
                }}
              >
                Checkout
              </Button>
            </Link>
          </NextLink>
        </Flex>
      </Flex>
    </Box>
  )
}

export default Bag
