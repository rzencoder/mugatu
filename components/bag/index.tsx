import { useBag } from '@/context/bagContext'
import { calculateSubTotal } from '@/utils/bag'
import { Flex, Box, useColorMode } from '@chakra-ui/react'
import BagItem from './bagItem'
import EmptyBag from './emptyBag'
import Summary from './summary'

const Bag = () => {
  const { bag } = useBag()
  const { colorMode } = useColorMode()
  const subtotal = calculateSubTotal(bag)

  if (!bag || bag.length === 0) {
    return <EmptyBag />
  }

  return (
    <Box bg={colorMode === 'light' ? '#eee' : 'mainBlack'}>
      <Flex
        direction={['column', null, 'row']}
        maxWidth="1000px"
        minHeight="500px"
        margin="0 auto"
        padding={['15px 5px', '20px']}
        justifyContent="space-between"
        position="relative"
      >
        <Flex width={['100%', null, '60%']} direction="column" mr={['5px', '20px']}>
          <Flex
            direction="column"
            p={['15px 10px', '20px']}
            mb="20px"
            bg={colorMode === 'light' ? 'mainWhite' : '#111'}
          >
            <Box p="0 0 15px" fontWeight="600" fontSize="28px">
              my bag
            </Box>
            {bag.map((item) => {
              return <BagItem key={`bag-item-${item.id}`} item={item} />
            })}
          </Flex>
          <Flex
            display={['none', null, 'flex']}
            justifyContent="space-between"
            bg={colorMode === 'light' ? 'mainWhite' : '#111'}
            p="20px"
            fontSize="22px"
            fontWeight="600"
          >
            <Box>sub-total</Box>
            <Box>Â£{subtotal}</Box>
          </Flex>
        </Flex>
        <Summary subtotal={subtotal} />
      </Flex>
    </Box>
  )
}

export default Bag
