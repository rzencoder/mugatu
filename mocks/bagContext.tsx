import { bagContext } from '@/context/bagContext'
import { BagItem } from '@/types/bagItem'
import { useState } from 'react'

interface Result {
  title: string
  message: string
  status: string
}

interface BagContextInterface {
  bag: BagItem[]
  fetchBag: () => Promise<void>
  fetchBagOnSignIn: () => Promise<void>
  addToBag: (item: BagItem, signedIn?: boolean) => Promise<Result>
  removeFromBag: (item: BagItem, signedIn?: boolean) => Promise<Result>
}

interface ProvideBagProps {
  bag: BagItem[]
  children: React.ReactNode
}

export const ProvideBag = ({ bag, children }: ProvideBagProps): JSX.Element => {
  const bagValue = useProvideBag(bag)
  return <bagContext.Provider value={bagValue}>{children}</bagContext.Provider>
}

export const useProvideBag = (mockBag: BagItem[]): BagContextInterface => {
  const [bag, setBag] = useState(mockBag)

  const fetchBag = async () => {
    try {
      setBag([...mockBag])
    } catch (error) {
      console.log(error)
    }
  }

  const fetchBagOnSignIn = async () => {
    if (bag.length > 0) {
      // Combine guest shopping bag with users saved bag
      try {
        setBag([...mockBag])
      } catch {
        console.log('there was an error fetching your shopping bag')
      }
    } else {
      try {
        await fetchBag()
      } catch {
        console.log('there was an error fetching your shopping bag')
      }
    }
  }

  const addToBag = async (item: BagItem, signedIn = false) => {
    if (signedIn) {
      try {
        setBag([...bag, item])
        return {
          title: 'Item Added!',
          message: 'Your item has been added to your shopping bag',
          status: 'success',
        }
      } catch {
        return {
          title: 'Error!',
          message: 'There was an error adding your item',
          status: 'error',
        }
      }
    } else {
      const filteredBag = bag.filter((product: BagItem) => product.id !== item.id)
      setBag([...filteredBag, item])
      return {
        title: 'Item Added!',
        message: 'Your item has been added to your shopping bag',
        status: 'success',
      }
    }
  }

  const removeFromBag = async (item: BagItem, signedIn = false) => {
    if (!signedIn) {
      const updatedBag = bag.filter((el: BagItem) => el.id !== item.id)
      setBag(updatedBag)
      return {
        title: 'Item Removed!',
        message: 'Your item has been removed from your shopping bag',
        status: 'success',
      }
    } else {
      try {
        const updatedBag = bag.filter((el: BagItem) => el.id !== item.id)
        setBag(updatedBag)
        return {
          title: 'Item Removed!',
          message: 'Your item has been removed',
          status: 'success',
        }
      } catch {
        return {
          title: 'Error!',
          message: 'There was an error removing your item',
          status: 'error',
        }
      }
    }
  }

  return {
    bag,
    fetchBag,
    fetchBagOnSignIn,
    addToBag,
    removeFromBag,
  } as const
}
