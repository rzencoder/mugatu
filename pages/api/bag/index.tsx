import { addToBag, getBag, removeFromBag } from '@/controllers/bag'
import { NextApiRequest, NextApiResponse } from 'next'

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export default async (req: NextApiRequest, res: NextApiResponse) => {
  const item = req.body
  switch (req.method) {
    case 'GET':
      console.log('here')
      return getBag(req, res)
    case 'POST':
      return addToBag(req, res, item)
    case 'PUT':
      return removeFromBag(req, res, item)
    default:
      //Method not allowed
      return res.status(405).end()
  }
}
