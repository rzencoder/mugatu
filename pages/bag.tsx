import Head from 'next/head'
import { Layout } from '@/components/layouts'
import Bag from '@/components/bag'

export default function BagPage(): JSX.Element {
  return (
    <>
      <Head>
        <title>Shopping Bag | Mugatu</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <Layout>
        <Bag />
      </Layout>
    </>
  )
}
