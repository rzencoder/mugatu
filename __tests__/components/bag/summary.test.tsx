import { fireEvent, render } from '@testing-library/react'
import Summary from '@/components/bag/summary'
import { ProvideBag } from '@/mocks/bagContext'
import { bag } from '@/mocks/bag'

describe('<Summary />', () => {
  it('renders <Summary /> component', () => {
    const { container, getByText } = render(
      <ProvideBag bag={[]}>
        <Summary subtotal={49.99} />
      </ProvideBag>
    )
    expect(getByText('£49.99')).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('changes delivery prices when subtotal is above £50', () => {
    const { getAllByText, getByTestId } = render(
      <ProvideBag bag={bag}>
        <Summary subtotal={59.99} />
      </ProvideBag>
    )
    fireEvent.click(getByTestId('set-delivery-standard delivery'))
    expect(getAllByText('standard delivery (free)')).toBeTruthy()
  })
})
