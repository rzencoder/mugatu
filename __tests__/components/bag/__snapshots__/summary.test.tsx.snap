// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Summary /> renders <Summary /> component 1`] = `
<div
  class="css-1ag7b0c"
>
  <div
    class="css-14r8l3o"
  >
    summary
  </div>
  <div
    class="css-1h1ghi2"
  >
    <div
      class="css-1ydx9ur"
    >
      <div
        class="css-0"
      >
        sub-total
      </div>
      <div
        class="css-0"
      >
        £
        49.99
      </div>
    </div>
    <div
      class="css-19qn5v4"
    >
      <svg
        class="chakra-icon css-1w891y"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
          fill="currentColor"
        />
      </svg>
      <div
        class="css-0"
      >
        spend £0.01 more to qualify for free delivery
      </div>
    </div>
    <div
      class="css-j7qwjs"
    >
      <div
        class="css-1ydx9ur"
      >
        <div
          class="css-0"
        >
          delivery
        </div>
        <div
          class="css-0"
        >
          £4.95
        </div>
      </div>
      <button
        aria-controls="menu-list-6"
        aria-expanded="false"
        aria-haspopup="menu"
        class="chakra-button chakra-menu__menu-button css-1k62m3q"
        id="menu-button-6"
        type="button"
      >
        <span
          class="css-a9hmjk"
        >
          <span
            class="css-0"
          >
            standard delivery (£4.95)
          </span>
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
      <div
        class="css-0"
        style="position: absolute; left: 0px; top: 0px; visibility: hidden;"
      >
        <div
          aria-orientation="vertical"
          class="chakra-menu__menu-list css-18un4sm"
          id="menu-list-6"
          role="menu"
          style="transform-origin: top left; opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
          tabindex="-1"
        >
          <button
            class="chakra-menu__menuitem css-4dys5b"
            data-testid="set-delivery-standard delivery"
            id="menuitem-1"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            standard delivery (£4.95)
          </button>
          <button
            class="chakra-menu__menuitem css-4dys5b"
            data-testid="set-delivery-next day delivery"
            id="menuitem-2"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            next day delivery (£6.95)
          </button>
          <button
            class="chakra-menu__menuitem css-4dys5b"
            data-testid="set-delivery-click and collect standard"
            id="menuitem-3"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            click and collect standard (£4.95)
          </button>
          <button
            class="chakra-menu__menuitem css-4dys5b"
            data-testid="set-delivery-click and collect next day"
            id="menuitem-4"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            click and collect next day (£6.95)
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-1ps467w"
    >
      <div
        class="css-0"
      >
        total
      </div>
      <div
        class="css-0"
      >
        £
        54.94
      </div>
    </div>
  </div>
  <a
    class="chakra-link css-0"
    href="/checkout"
  >
    <button
      class="chakra-button css-1djzb9a"
      type="button"
    >
      Checkout
    </button>
  </a>
</div>
`;
